queasars/utility/spsa_termination.py

```diff
@@ -0,0 +1,143 @@
+    absolute difference between the last and new function value, divided by the last function value,
+    does not fall below minimum_relative_change for more than allowed_consecutive_violations times.
+    This termination checker also keeps track of the best function value and parameter values
+    seen during the optimization, as well as the entire history of observed function values.
+
+    :param minimum_relative_change: threshold value for the change in function value relative to the last
+        function value, below which this termination checker chooses to terminate the SPSA optimization
+    :type minimum_relative_change: float
+    :param allowed_consecutive_violations: determines how often the threshold value can be violated consecutively
+        before this termination checker chooses to terminate. If set to 0, this terminates the first time the change
+        falls below the threshold value. If set to 2 for example, this terminates the first time the change falls below
+        the threshold three consecutive times. Must be at least 0.
+    :type allowed_consecutive_violations: int
+    :param maxfev: maximum amount of function evaluations the SPSA optimizer may utilize
+    :type maxfev: Optional[int]
+    """
+
+    def __init__(
+        self,
+        minimum_relative_change: float,
+        allowed_consecutive_violations: int,
+        maxfev: Optional[int] = None,
+    ):
+        self._minimum_relative_change: float = minimum_relative_change
+        self._allowed_consecutive_violations: int = allowed_consecutive_violations
+        self._maxfev: Optional[int] = maxfev
+        self._function_value_history: list[float] = []
+        self._change_history: list[float] = []
+        self._n_function_evaluations = 0
+        self._n_function_evaluation_history: list[float] = []
+        self._best_function_value: float = float("inf")
+        self._best_parameter_values: Optional[NDArray] = None
+        self._done: bool = False
+
+    def termination_check(
+        self,
+        n_function_evaluations: int,
+        parameter_values: NDArray,
+        function_value: float,
+        step_size: float,
+        accepted: bool,
+    ) -> bool:
+        """Given the callback values provided by qiskit_algorithm's SPSA optimizer, this method determines
+        whether the SPSA optimization should terminate"""
+
+        if self._done:
+            self._function_value_history = []
+            self._change_history = []
+            self._n_function_evaluations = 0
+            self._n_function_evaluation_history = []
+            self._best_function_value = float("inf")
+            self._best_parameter_values = None
+            self._done = False
+
+        self._n_function_evaluations = n_function_evaluations
+
+        if self._maxfev is not None and self._n_function_evaluations >= self._maxfev:
+            return True
+
+        if not accepted:
+            return False
```

Comment:
What does `accepted` mean as an argument for the call back?

Reply:
There is a parameter called `blocking` for spsa that after a gradient step evaluates the new parameters and rejects the update step if it worsens the function value. Accepted specifies whether a reported step was accepted by SPSA when using blocking. If an update step was rejected, checking against its stability for convergence would be incorrect as far as i would understand it.

Reply:
Ah ok, right, thank you, you explained that SPSA feature to me some time ago, I didn't notice this was related to it.